Features Gater effect #1299026

johanLsp opened the pull request:
This implements a Gater effect as discussed in [`](https://bugs.launchpad.net/mixxx/+bug/1299026)

The gate is shaped with an attack, a hold and a release phase. The hold time is derived from the Shape parameter.

The gate is triggered according to the Rate parameter, triggers happen at subdivision of the beat if a beat grid is available, or periodically otherwise.
This effect maintains a per channel state, meaning that two channels sharing the same effect slots won't interfere with each other.

johanLsp committed:
Add dummy Gater effect

johanLsp committed:
Simple working Gater w/ Rate & Shape params

johanLsp committed:
Add Shape parameter

johanLsp committed:
Add ADSR envelope gating

johanLsp committed:
Implements a triggered Attack-Hold-Release gate

daschuer commented on src/effects/native/gatereffect.cpp:
```diff
@@ -0,0 +1,220 @@
+    const auto shape  = m_pShapeParameter->value();
+
+    // Used to divide a beat by a power of 2
+    int divider = pow(2, ((int)(3*rate)));
```
you can use floor() her to avoid the integer double roundtrip

daschuer commented on src/effects/native/gatereffect.cpp:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    // Write back channel state
+    pState->timePosition[handle] = timePosition;
```
There is no need to add [handle] because there is already passed the handle's pState.

daschuer commented on src/effects/native/gatereffect.h:
```diff
@@ -0,0 +1,67 @@
+
+struct GaterGroupState {
+    enum State {IDLE, ATTACK, HOLD, RELEASE};
+    ChannelHandleMap<State> state;
```
no need for a map here

daschuer commented on src/effects/native/gatereffect.h:
```diff
@@ -0,0 +1,67 @@
+
+    // Fixed parameters
+    // Gain when the gate is open, higher than 1 to still have a decent level
+    double maxGain    = 1.5; 
```
We do not add initializers to class definition. 
Please put these constants into an anonymous namespace in the cpp file 
Like this 

```c++
namespace {
constexpr double kBaseAttackInc = 0.001;
...
} // anonymous namespace 
```

daschuer reviewed (changes requested):
Thank you for the nice work. I have added some minor comments. 

There are also some code style issues. If you like, you may fix them according to: https://www.mixxx.org/wiki/doku.php/coding_guidelines

daschuer commented:
Before merge, we need your permission. 
Please sign: https://docs.google.com/a/mixxx.org/spreadsheet/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ

and comment here when done. Thank you.

johanLsp committed:
Conform to code guidelines

johanLsp committed:
Remove unnecessary handle states

johanLsp commented:
Thanks for the feedback, I signed the google doc and went through the coding guidelines, it should be fine now if I didn't miss anything

daschuer commented:
Thank you for the updates. I have just played with it and it is already a nice addition. 
Here are my findings:
* The Invert knob is a nice idea. Can you replace it with a phase knob, that allows to highlight any place in a measure? It should shift the effect start between the beats. 
* If you turn the rate knob, the effect falls out of phase. You get short cycles even if you are on the long scale side. 
* The tripplets knob at the Echo speeds up the effect by 3. This is a nice "effect" and we should adopt this here as well. 
* The rate knob should be continuously, not stepping. This will allow to smooth adopt the effect to a track without a beat-grid. You may adopt the Echo Time knob. Or you may adopt the speed knob from the upcoming flanger, after it is merged https://github.com/mixxxdj/mixxx/pull/1362

uklotzde commented on src/effects/native/gatereffect.cpp:
```diff
@@ -0,0 +1,236 @@
+#include "effects/native/gatereffect.h"
+
+namespace {
+    // Gain when the gate is open, higher than 1 to still have a decent level
```
We omit indentation within (possibly nested) namespaces to keep the code readable.

uklotzde commented on src/effects/native/gatereffect.cpp:
```diff
@@ -0,0 +1,236 @@
+            timePosition = groupFeatures.beat_fraction*beatPeriod;
+        } else {
+            timePosition = 0;   
+            beatPeriod   = sampleRate; // 1 second
```
Correct, this is 1 second in _sample frames_, but not when counting individual samples like the engine does. The engine design is somehow broken by indexing total number of samples from (implicitly) 2 channels instead of sample frames with interleaved samples. Indexing individual samples is dependent on the number of channels, actually 2. This design prevents processing of multiple channels or optimizations for mono signals.

It is not clear to me where and for which of the parameters you implicitly take this magic number of 2 channels into account. Maybe it would help to explicitly define a constexpr kNumberOfChannels = 2 and use it where applicable. In the next step you can rewrite your code to minimize usage of this constant. It should only be needed initially to transform the input parameters (divide by kNumberOfChannels) and finally when you actually need a "sample index" (multiply by kNumberOfChannels, adding 1 for the second channel).

If you try to keep your calculations independent of the number of channels and already use "sample frames" for the internal calculations your code becomes more obvious and migration will be easier, once the engine interfaces have been fixed.

Not your fault, this is a legacy issue ;)

uklotzde replied:
Consequently all calculations should be dependent only on the sample rate, but not on the number of channels. Multiplication/division by 2 is fast anyway, so I would recommend to strive for clear code instead of maximizing the performance for a broken interface design.

uklotzde commented on src/effects/native/gatereffect.cpp:
```diff
@@ -0,0 +1,236 @@
+            // Hold : Hold maxGain for holdCounter samples
+            case GaterGroupState::HOLD:
+                gain = kMaxGain;
+                if(holdCounter == 0) {
```
"if" is a keyword and not a function. Please insert a space between the keyword and the opening brace to improve readability. You can fix this when you need to edit the code anyway. Formatting the code according to syntactical rules should better be done with an automatic code formatter, that we don't have ;)

uklotzde commented on src/effects/native/gatereffect.h:
```diff
@@ -0,0 +1,51 @@
+
+
+struct GaterGroupState {
+    enum State {IDLE, ATTACK, HOLD, RELEASE} state;
```
Use C++ "enum class" to improve type safety and limit visibility. The latter is not an issue here, but might be elsewhere.

uklotzde commented on src/effects/native/gatereffect.h:
```diff
@@ -0,0 +1,51 @@
+class GaterEffect : public PerChannelEffectProcessor<GaterGroupState> {
+  public:
+    GaterEffect(EngineEffect* pEffect, const EffectManifest& manifest);
+    virtual ~GaterEffect();
```
Replace "virtual" with "override" for improved compile time checks:
```
~GaterEffect() override;
```

uklotzde commented on src/effects/native/gatereffect.h:
```diff
@@ -0,0 +1,51 @@
+                        const unsigned int numSamples,
+                        const unsigned int sampleRate,
+                        const EffectProcessor::EnableState enableState,
+                        const GroupFeatureState& groupFeatures);
```
Add "override" for improved compile time checks:
```
...) override;
```

Be-ing commented on src/effects/native/gatereffect.cpp:
```diff
@@ -0,0 +1,236 @@
+
+namespace {
+    // Gain when the gate is open, higher than 1 to still have a decent level
+    constexpr double kMaxGain        = 1.5; 
```
I don't understand why this isn't 1. It is unintuitive that this effect would add any gain, at least by default. It makes it fairly likely that a deck that is playing at a good level will clip when this effect is turned on. It also makes it impossible to have a smooth transition between turning the effect on and off because there is a sudden volume jump.

Be-ing commented on src/effects/native/gatereffect.cpp:
```diff
@@ -0,0 +1,236 @@
+    manifest.setName(QObject::tr("Gater"));
+    manifest.setAuthor("The Mixxx Team");
+    manifest.setVersion("1.0");
+    manifest.setDescription("");
```
Please add a description for the effect and every parameter to describe what they do and how the scale of the parameter works. The descriptions appear as tooltips (when tooltips are enabled).

Be-ing replied:
Also add a shortName for the effect and every parameter, even if it is identical to the full name in English so it can be translated.

Be-ing commented on src/effects/native/gatereffect.cpp:
```diff
@@ -0,0 +1,236 @@
+
+    EffectManifestParameter* shape = manifest.addParameter();
+    shape->setId("shape");
+    shape->setName(QObject::tr("Shape"));
```
Maybe "Hold" would be a better name for this parameter?

johanLsp replied:
I began to change it to "Hold" after your comment, but I think that "Shape" is appropriate and will fit better when moving to the more generic Tremolo effect discussed below

Be-ing commented on src/effects/native/gatereffect.cpp:
```diff
@@ -0,0 +1,236 @@
+    const double shape  = m_pShapeParameter->value();
+
+    // Used to divide a beat by a power of 2
+    int divider = pow(2, floor(3*rate));
```
I don't understand this scaling. It feels random. Could you explain your reasoning for this?

johanLsp replied:
The aim here is to have a "round" division of beats, that is, an effect period of 1/8, 1/4, 1/2 or 1 beat (when quantization is on of course), as this keeps the effect aligned with beats.
This divider is bypassed when quantization is off

Be-ing commented on src/effects/native/gatereffect.cpp:
```diff
@@ -0,0 +1,236 @@
+        triggerTime = 0;
+    }
+
+    // holdTime : number of samples spent in the Hold state
```
Use variable names to clearly differentiate variables that are in units of samples versus time (seconds). For example, `holdTimeSamples` or `holdSamples`.

Be-ing commented on src/effects/native/gatereffect.cpp:
```diff
@@ -0,0 +1,236 @@
+            // Attack : Increase gain up to maxGain, then Hold
+            case GaterGroupState::ATTACK:
+                gain += attackInc;
+                if(gain >= kMaxGain) {
```
space between `if` and `(`

Be-ing commented on src/effects/native/gatereffect.cpp:
```diff
@@ -0,0 +1,236 @@
+    rate->setId("rate");
+    rate->setName(QObject::tr("Rate"));
+    rate->setDescription(QObject::tr(""));
+    rate->setControlHint(EffectManifestParameter::ControlHint::KNOB_STEPPING);
```
Hmm, should this really be stepping? With the Echo effect, it only acts as a stepping knob when the Quantize button parameter is enabled. Is there a use case for not rounding this to a specific fraction of a beat and sweeping it across a continuous range?

Be-ing commented on src/effects/native/gatereffect.cpp:
```diff
@@ -0,0 +1,236 @@
+    rate->setDescription(QObject::tr(""));
+    rate->setControlHint(EffectManifestParameter::ControlHint::KNOB_STEPPING);
+    rate->setSemanticHint(EffectManifestParameter::SemanticHint::UNKNOWN);
+    rate->setUnitsHint(EffectManifestParameter::UnitsHint::UNKNOWN);
```
This parameter should be linked to the metaknob by default.

Be-ing reviewed (changes requested):
Thanks for working on this! It is a good start.

Be-ing commented on src/effects/native/gatereffect.cpp:
```diff
@@ -0,0 +1,236 @@
+    rate->setControlHint(EffectManifestParameter::ControlHint::KNOB_STEPPING);
+    rate->setSemanticHint(EffectManifestParameter::SemanticHint::UNKNOWN);
+    rate->setUnitsHint(EffectManifestParameter::UnitsHint::UNKNOWN);
+    rate->appendStep(QPair<QString, double>("1", 0));
```
Maybe the maximum should be above 1 beat. How about 2 beats like Echo?

johanLsp committed:
Unindent anonymous namespace

johanLsp committed:
Set Rate knob to continuous

johanLsp committed:
Fix phase alignment when moving manually in track

johanLsp committed:
Change Triplet toggle speed

johanLsp committed:
Process independently of the number of channels

johanLsp committed:
Switch from Invert to more generic Phase parameter

daschuer commented:
I just remember the Goaslicer: the https://github.com/mixxxdj/mixxx/pull/337 which seams to be essentially the same. 

Reading this, I am starting to struggling with the name. For all our effects there is a stomp-box equivalent implemented in hardware. But there is no Gater stomp-box. It looks like this is a Traktor offspring. 
There is are plenty "Noise Gater" stomp-boxes which are different effect.

Then I have found the Tremolo effect. https://en.wikipedia.org/wiki/Tremolo 
And I think this is the one. Right? 

https://www.google.de/search?q=tremolo+effect&dcr=0&source=lnms&tbm=isch&sa=X&ved=0ahUKEwiU9NXXp6jXAhXIpaQKHeANAAoQ_AUIDCgD&biw=1301&bih=637
 
How about to rename it to Tremolo?

johanLsp commented:
This definitely looks like a tremolo effect, or at least a specific use of it
Stompboxes seem to include the following parameters:

- Rate : Obviously the same as currently implemented
- Shape/Waveform : Choose between a sine, sawtooth, square waveform.
- Symmetry : What I implemented as the "Shape" parameter, linked to the duty cycle of the waveform
- Depth : I think the effects' "Mix" knob is doing just that by weighting the gain

I already added a "Phase" parameter as you mentionned (not commited yet), so Rate, Shape, Symmetry & Phase could be the 4 parameters of this more generic Tremolo effect.
The current implementation is too specific though and calling it a tremolo might be confusing. What I can do is try and play on another branch on the tremolo effect, and close this pull request if the results are promising.

johanLsp committed:
Renamed to Tremolo

Be-ing commented:
Ping @johanLsp, any thoughts about my review comments above? We'd like to have this fun effect merged for 2.1.

johanLsp commented:
I should have some time by the end of the week to take into account the last changes needed.
As daschuer suggested, the effect is renamed tremolo and will stay as is, additional features will be implemented later on.

johanLsp committed:
Add description

johanLsp commented:
Pushed the last commits addressing the comments above.
I tried to address all the code style issues, further functionalities to implement a stompbox-like tremolo will be part of a future pull request

johanLsp committed:
Implement Tremolo features

johanLsp commented:
This commit introduces the tremolo specific features discussed earlier.
Phase, Quantize & Triplet parameters have the same effect.
**Shape** refers to what was also described as symmetry, which is the duty cycle of signal modulating the amplitude
**Smooth** allows to progressively transition from a square wave modulation to a sine wave modulation
**Rate** : the period of the effect can now be set to up to 4 beats, maximum rate still being 1/8th of a beat (or second)
The implementation is now quite different from what I started with, but this starts to feel and sound good.

Be-ing commented:
I gave this a quick test and it's fun! Thank you! I put it second in a chain with Echo after it and played with loading various effects first in the chain. Have you come up with any ways to use this in chains?

I'll do a more thorough review at a later time, but a few issues jumped out at me:
  * Parameter descriptions need "\n" for line breaks.
  * I don't really understand what the Shape parameter does or its description. Try to write the description in language that would be understandable to a DJ who has no idea how the effect works inside.

johanLsp commented:
Any update on this ?

daschuer commented on src/effects/native/tremoloeffect.cpp:
```diff
@@ -0,0 +1,224 @@
+
+    EffectManifestParameter* shape = manifest.addParameter();
+    shape->setId("shape");
+    shape->setName(QObject::tr("Shape"));
```
For shape I expect something like rectangular or sin wave. How do stop box manufacturer think about that?

daschuer commented on src/effects/native/tremoloeffect.cpp:
```diff
@@ -0,0 +1,224 @@
+    shape->setMaximum(0.9);
+
+    EffectManifestParameter* smooth = manifest.addParameter();
+    smooth->setId("smooth");
```
I think this is normally called "wave".
If you have an other hardware model with other names it works also me. I just want to make sure we do not create our own namespace.

johanLsp replied:
Most stompboxes come with a small drawing for what I called "shape" & "smooth" parameter
We can use what the Shapeshifter from seymour duncan uses : [Shapeshifter](https://www.google.fr/search?q=shapeshifter+seymour+duncan&client=ubuntu&channel=fs&dcr=0&source=lnms&tbm=isch&sa=X&ved=0ahUKEwj_gczdgPvXAhXCIlAKHZ6xBdkQ_AUICygC&biw=1533&bih=769#imgrc=DQUxHyxZeMJpgM:)
- "Smooth" would be called Wave (or Waveform, more explicit without any drawing)
- "Shape" would stay the same

daschuer replied:
OK

daschuer commented on src/effects/native/tremoloeffect.cpp:
```diff
@@ -0,0 +1,224 @@
+    const double smooth = m_pSmoothParameter->value();
+
+    unsigned int currentFrame = pState->currentFrame;
+    double       gain         = pState->gain;
```
nit: we do not care about alignment of code lines.
This will likely brake if you refactor tho code.
Use just single spaces.

daschuer commented on src/effects/native/tremoloeffect.cpp:
```diff
@@ -0,0 +1,224 @@
+    unsigned int currentFrame = pState->currentFrame;
+    double       gain         = pState->gain;
+
+    const GroupFeatureState& gf = groupFeatures;
```
Why you need a local copy of the reference?
This might be confusing in future, because they reference the same memory region.

johanLsp replied:
This was intended as an alias to improve code readability as the original name was too long. I didn't want to change the parameter name in the function declaration directly to keep the meaning of the variable displayed somewhere. We can still revert to using the full name directly

daschuer reviewed (commented):
Good work! I have not yet tested it, but from the code it looks good. I have added only some style an naming nits.

johanLsp committed:
Removed additional spaces

daschuer commented:
Manual test past as well.Nice :-)

johanLsp committed:
Change parameter names and description

Be-ing commented on src/effects/native/tremoloeffect.cpp:
```diff
@@ -51,27 +51,27 @@ EffectManifest TremoloEffect::getManifest() {
+    waveform->setName(QObject::tr("Waveform"));
+    waveform->setShortName(QObject::tr("Waveform"));
+    waveform->setDescription(QObject::tr("Sets the waveform of the modulation\n"
+    "0 - Square wave\n"
```
Users can't explicitly set the numeric value of parameters, so use "Fully left" and "Fully right" instead of "0" and "1".

Be-ing commented on src/effects/native/tremoloeffect.cpp:
```diff
@@ -42,7 +42,7 @@ EffectManifest TremoloEffect::getManifest() {
     shape->setName(QObject::tr("Shape"));
     shape->setShortName(QObject::tr("Shape"));
-    shape->setDescription(QObject::tr("Sets the duty cycle of the modulation"
+    shape->setDescription(QObject::tr("Sets the duty cycle of the modulation\n"
```
I don't know what this means and neither will many users.

Be-ing reviewed (changes requested):
Sorry for the delay reviewing this. Here are a few more comments.

johanLsp committed:
Update parameters description

daschuer commented:
LGTM @Be-ing merge?

Be-ing commented:
I'll give it another test tomorrow.

Be-ing committed:
explicitly cast int to double

Be-ing committed:
fix indentation

Be-ing committed:
add a Depth parameter to Tremolo and link it to metaknob

This allows for smoothly bringing the effect in and out of a mix more
easily. Previously the effect had to be disabled exactly at the peak
volume to turn it off smoothly.

Be-ing commented:
I am still quite confused by the Shape parameter. It sounds like the higher the parameter is set, the longer the peak of the volume is. Am I understanding that correctly? Maybe "Width" would be a better name?

I have sent you a pull request to add a Depth parameter to make it easier to bring this effect in and out of a mix.

Be-ing commented:
Hi @johanLsp, we are releasing 2.1 beta this Friday 2017-12-22. Can we come to a conclusion about the parameter names and descriptions and get this merged by then?

johanLsp commented:
I'll do this today, width seems a good name to me

johanLsp committed:
Merge pull request #1 from Be-ing/tremolo

Tremolo Depth parameter

johanLsp committed:
Change Shape parameter name to Width

Be-ing commented:
Sounds good to me! 👍  Thank you for this fun effect! I'll take care of adapting it for the effect API changes in #1254 once this is merged.

daschuer commented:
LGTM thank you very much!
